src/Jellyfin.MediaEncoding.Hls/ScheduledTasks/KeyframeExtractionScheduledTask.cs

```diff
@@ -0,0 +1,100 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Jellyfin.Data.Enums;
+using Jellyfin.MediaEncoding.Hls.Extractors;
+using MediaBrowser.Controller.Dto;
+using MediaBrowser.Controller.Entities;
+using MediaBrowser.Controller.Library;
+using MediaBrowser.Model.Entities;
+using MediaBrowser.Model.Globalization;
+using MediaBrowser.Model.Tasks;
+
+namespace Jellyfin.MediaEncoding.Hls.ScheduledTasks;
+
+/// <inheritdoc />
+public class KeyframeExtractionScheduledTask : IScheduledTask
+{
+    private readonly ILocalizationManager _localizationManager;
+    private readonly ILibraryManager _libraryManager;
+    private readonly IKeyframeExtractor[] _keyframeExtractors;
+    private static readonly BaseItemKind[] _itemTypes = { BaseItemKind.Episode, BaseItemKind.Movie };
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="KeyframeExtractionScheduledTask"/> class.
+    /// </summary>
+    /// <param name="localizationManager">An instance of the <see cref="ILocalizationManager"/> interface.</param>
+    /// <param name="libraryManager">An instance of the <see cref="ILibraryManager"/> interface.</param>
+    /// <param name="keyframeExtractors">The keyframe extractors.</param>
+    public KeyframeExtractionScheduledTask(ILocalizationManager localizationManager, ILibraryManager libraryManager, IEnumerable<IKeyframeExtractor> keyframeExtractors)
+    {
+        _localizationManager = localizationManager;
+        _libraryManager = libraryManager;
+        _keyframeExtractors = keyframeExtractors.ToArray();
+    }
+
+    /// <inheritdoc />
+    public string Name => "Keyframe Extractor";
+
+    /// <inheritdoc />
+    public string Key => "KeyframeExtraction";
+
+    /// <inheritdoc />
+    public string Description => "Extracts keyframes from video files to create more precise HLS playlists";
+
+    /// <inheritdoc />
+    public string Category => _localizationManager.GetLocalizedString("TasksLibraryCategory");
+
+    /// <inheritdoc />
+    public Task Execute(CancellationToken cancellationToken, IProgress<double> progress)
+    {
+        var query = new InternalItemsQuery
+        {
+            MediaTypes = new[] { MediaType.Video },
+            IsVirtualItem = false,
+            IncludeItemTypes = _itemTypes,
+            DtoOptions = new DtoOptions(true),
+            SourceTypes = new[] { SourceType.Library },
+            Recursive = true
+        };
+
+        var videos = _libraryManager.GetItemList(query);
+        var numberOfVideos = videos.Count;
+        var numComplete = 0;
+
+        // TODO parallelize with Parallel.ForEach?
```

Comment:
Not without a setting I think, it'll kill either IO performance of the CPU (or both).